File: src/main/java/cn/enaium/epsilon/mixin/KeyboardMixin.java
package cn.enaium.epsilon.mixin;

import cn.enaium.cf4m.CF4M;
import cn.enaium.epsilon.client.events.KeyboardEvent;
import net.minecraft.client.Keyboard;
import net.minecraft.client.MinecraftClient;
import org.lwjgl.glfw.GLFW;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

/**
 * Project: Epsilon
 * -----------------------------------------------------------
 * Copyright © 2020 | Enaium | All rights reserved.
 */
@Mixin(Keyboard.class)
class KeyboardMixin {
    @Inject(at = @At("HEAD"), method = "onKey")
    private void onKey(long windowHandle, int keyCode, int scanCode, int action, int modifiers, CallbackInfo callbackInfo) {
        if (action == GLFW.GLFW_PRESS && MinecraftClient.getInstance().currentScreen == null) {
            CF4M.module.onKey(keyCode);
            new KeyboardEvent(keyCode).call();
        }
    }
}

File: src/main/kotlin/cn/enaium/epsilon/mixin/KeyboardMixin.kt
package cn.enaium.epsilon.mixin

import cn.enaium.epsilon.event.events.EventKeyboard
import net.minecraft.client.Keyboard
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

/**
 * Project: Epsilon
 * -----------------------------------------------------------
 * Copyright © 2020 | Enaium | All rights reserved.
 */
@Mixin(Keyboard::class)
class KeyboardMixin {
    @Inject(at = [At("HEAD")], method = ["onKey"])
    private fun onKey(windowHandle: Long, keyCode: Int, scanCode: Int, action: Int, modifiers: Int, callbackInfo: CallbackInfo) {
        EventKeyboard(keyCode, scanCode, action, modifiers).call()
    }
}